src/Cooked/MockChain/GenerateTx.hs

```diff
@@ -226,15 +217,48 @@ txOutRefsToTxInsReference =
     )
     . mapM Ledger.toCardanoTxIn
 
--- Convert a list of 'Api.TxOutRef' into a 'Cardano.TxInsCollateral'
-txOutRefsToTxSkelInsCollateral :: [Api.TxOutRef] -> TxGen (Cardano.TxInsCollateral Cardano.ConwayEra)
-txOutRefsToTxSkelInsCollateral =
-  throwOnToCardanoError "txOutRefsToTxInCollateral"
-    . fmap toTxInsCollateral
-    . mapM Ledger.toCardanoTxIn
-  where
-    toTxInsCollateral [] = Cardano.TxInsCollateralNone
-    toTxInsCollateral ins = Cardano.TxInsCollateral Cardano.AlonzoEraOnwardsConway ins
+-- Computes the collateral triplet from the fee and the collateral inputs in
+-- the context
+toCollateralTriplet ::
+  TxGen
+    ( Cardano.TxInsCollateral Cardano.ConwayEra,
+      Cardano.TxTotalCollateral Cardano.ConwayEra,
+      Cardano.TxReturnCollateral Cardano.CtxTx Cardano.ConwayEra
+    )
+toCollateralTriplet = do
+  knownTxOuts <- asks managedTxOuts
+  returnCollateralWallet <- asks returnCollateralWallet
+  networkId <- asks (Emulator.pNetworkId . params)
+  collateralInsList <- asks (Set.toList . collateralIns)
+  collateralInsValue <-
+    let collateralInsResolved = mapMaybe (`Map.lookup` knownTxOuts) collateralInsList
+     in if length collateralInsResolved /= length collateralInsList
+          then throwOnString "toCollateralTriplet: unresolved txOutRefs"
+          else return $ mconcat (Api.txOutValue <$> collateralInsResolved)
+  collateralPercentage <- asks (toInteger . fromMaybe 100 . Cardano.protocolParamCollateralPercent . Emulator.pProtocolParams . params)
+  coinTotalCollateral <- asks (Emulator.Coin . (+ 1) . (`div` 100) . (* collateralPercentage) . fee)
+  let txTotalCollateral = Cardano.TxTotalCollateral Cardano.BabbageEraOnwardsConway coinTotalCollateral
+      returnCollateralValue = collateralInsValue <> PlutusTx.negate (toValue coinTotalCollateral)
```

Comment:
This is of course not supposed to happen but in theory, because of some error elsewhere in the code, `returnCollateralValue` could be a non empty negative value. Maybe we should add a guard and failure here with a message in case it happens, to detect an issue early on.

Reply:
This is a good point, I will do it.